### PR TITLE
[PHP7.1] Strict and Scalar type hints setup via EasyCodingStandard

### PR DESCRIPTION
**Pull request by TomasVotruba**
- apply strict_types
- apply scalar types
- apply private constants
- add EasyCodingStandard
- add to CI

## Finished Sections

- [x] Configuration
- [x] Console
- [x] Generator
- [x] Parser
- [x] Reflection
- [x] Templating
- [x] Theme
- [x] Utils

## Changes

- drop memory limit, set to -1 would be fine in case of need
- drop question for override, already in configuration
- drop invalid elements - do not parse invalid code
- drop confusing theme option - use theme config instead
- drop reasons for invalid elements
- drop usedby artificial annotation
- drop shortcuts in command, were creating hidden knowledge
- **explicit parse exception message added**
  
image

### CODE REVIEW

**Comment by jankal** [comment]
@TomasVotruba hard at work!

**Comment by TomasVotruba** [comment]
@jankal You bet. Almost there :)

File: src/Console/Helper/ProgressBar.php [hunk 1]
```diff
@@ -53,10 +47,7 @@ public function increment($increment = 1)
     }
 
 
-    /**
-     * @return string
-     */
-    private function getBarFormat()
+    private function getBarFormat(): string
```
Comment (jankal) [inline src/Console/Helper/ProgressBar.php]: Typed return. Docblock?

File: src/Console/IO/IO.php [hunk 2]
```diff
@@ -22,59 +19,28 @@ class IO implements IOInterface
...
     private $output;
 
-    /**
-     * @var HelperSet
-     */
-    private $helperSet;
 
-
-    public function __construct(HelperSet $helperSet, InputInterface $input, OutputInterface $output)
+    public function __construct(InputInterface $input, OutputInterface $output)
```
Comment (jankal) [inline src/Console/IO/IO.php]: I'd suggest adding a docblock.

File: src/Console/IO/IO.php [hunk 3]
```diff
@@ -22,59 +19,28 @@ class IO implements IOInterface
...
         $this->input = $input;
         $this->output = $output;
     }
 
 
-    /**
-     * {@inheritdoc}
-     */
-    public function getInput()
+    public function getInput(): InputInterface
```
Comment (jankal) [inline src/Console/IO/IO.php]: Typed return. What about docblock?

File: src/Console/IO/IO.php [hunk 4]
```diff
@@ -22,59 +19,28 @@ class IO implements IOInterface
...
     {
         return $this->input;
     }
 
 
-    /**
-     * {@inheritdoc}
-     */
-    public function getOutput()
+    public function getOutput(): OutputInterface
```
Comment (jankal) [inline src/Console/IO/IO.php]: Typed return. What about docblock?

File: src/Console/IO/IO.php [hunk 5]
```diff
@@ -22,59 +19,28 @@ class IO implements IOInterface
...
-    {
-        return $this->output->writeln($message);
-    }
-
-
-    /**
-     * {@inheritdoc}
-     */
-    public function ask($question, $default = null)
+    public function writeln(string $message): void
```
Comment (jankal) [inline src/Console/IO/IO.php]: Typed param. What about docblock?

File: src/Console/Input/LiberalFormatArgvInput.php [hunk 6]
```diff
@@ -1,22 +1,23 @@
...
+ * @todo put to standalone package
+ */
+final class LiberalFormatArgvInput extends ArgvInput
 {
-
-    /**
-     * @return array
-     */
-    public function getOptions()
+    public function getOptions(): array
```
Comment (jankal) [inline src/Console/Input/LiberalFormatArgvInput.php]: Typed return. What about docblock?

File: src/Console/Input/LiberalFormatArgvInput.php [hunk 7]
```diff
@@ -57,37 +60,37 @@ private function splitByComma($value)
...
         return $value;
     }
 
 
-    /**
-     * @param string $value
-     * @return bool
-     */
-    private function containsComma($value)
+    private function containsComma(string $value): bool
```
Comment (jankal) [inline src/Console/Input/LiberalFormatArgvInput.php]: Typed return and param. What about docblock?

File: src/Contracts/Configuration/ConfigurationInterface.php [hunk 8]
```diff
@@ -1,124 +1,98 @@
...
+     * @param mixed[] $options
+     * @return mixed[]
      */
-    public function resolveOptions(array $options);
+    public function resolveOptions(array $options): array;
 
 
     /**
-     * @param string $name
      * @return mixed|NULL
```
Comment (jankal) [inline src/Contracts/Configuration/ConfigurationInterface.php]: NULL in fat letters? [PSR 2](http://www.php-fig.org/psr/psr-2/#keywords-and-truefalsenull) states:

> The PHP constants true, false, and null MUST be in lower case.

File: src/Contracts/Configuration/ConfigurationInterface.php [hunk 9]
```diff
@@ -1,124 +1,98 @@
...
 
 
     /**
-     * Return name of main library
-     *
-     * @return string
+     * Get property/method visibility level (public, protected or private, in binary code).
      */
-    public function getMain();
+    public function getVisibilityLevel(): int;
```
Comment (jankal) [inline src/Contracts/Configuration/ConfigurationInterface.php]: Uhm there is a existing docblock but nothing about this `int` return in here...

File: src/Contracts/Configuration/ConfigurationInterface.php [hunk 10]
```diff
@@ -1,124 +1,98 @@
...
 
 
     /**
-     * Are PHP Core elements documented.
-     *
-     * @return bool
+     * Return name of main library
      */
-    public function isPhpCoreDocumented();
+    public function getMain(): string;
```
Comment (jankal) [inline src/Contracts/Configuration/ConfigurationInterface.php]: Uhm there is a existing docblock but nothing about this `string` return in here...

File: src/Contracts/Configuration/ConfigurationInterface.php [hunk 11]
```diff
@@ -1,124 +1,98 @@
...
 
 
     /**
      * Are elements marked as "@internal" documented.
-     *
-     * @return bool
      */
-    public function isInternalDocumented();
+    public function isInternalDocumented(): bool;
```
Comment (jankal) [inline src/Contracts/Configuration/ConfigurationInterface.php]: Uhm there is a existing docblock but nothing about this `bool` return in there...

File: src/Contracts/Configuration/ConfigurationInterface.php [hunk 12]
```diff
@@ -1,124 +1,98 @@
...
      */
-    public function getAnnotationGroups();
+    public function getAnnotationGroups(): array;
 
 
-    /**
-     * @return string
-     */
-    public function getDestination();
+    public function getDestination(): string;
```
Comment (jankal) [inline src/Contracts/Configuration/ConfigurationInterface.php]: No docblock for that method even if it has a typed return of `string`.

File: src/Contracts/Configuration/ConfigurationInterface.php [hunk 13]
```diff
@@ -1,124 +1,98 @@
...
 
 
     /**
      * Get title of the project.
-     *
-     * @return string
      */
-    public function getTitle();
+    public function getTitle(): string;
```
Comment (jankal) [inline src/Contracts/Configuration/ConfigurationInterface.php]: Uhm there is a existing docblock but nothing about this `string` return in there...

File: src/Contracts/Configuration/ConfigurationInterface.php [hunk 14]
```diff
@@ -1,124 +1,98 @@
...
 
 
     /**
      * Base url of application.
-     *
-     * @var string
      */
-    public function getBaseUrl();
+    public function getBaseUrl(): string;
```
Comment (jankal) [inline src/Contracts/Configuration/ConfigurationInterface.php]: Uhm there is a existing docblock but nothing about this `string` return in there...

File: src/Contracts/Configuration/ConfigurationInterface.php [hunk 15]
```diff
@@ -1,124 +1,98 @@
...
 
 
-    /**
-     * @return string
-     */
-    public function getGoogleCseId();
+    public function getGoogleCseId(): string;
```
Comment (jankal) [inline src/Contracts/Configuration/ConfigurationInterface.php]: Typed `string` return but no docblock.

File: src/Contracts/Configuration/ConfigurationInterface.php [hunk 16]
```diff
@@ -1,124 +1,98 @@
...
 
 
-    /**
-     * @return bool
-     */
-    public function shouldGenerateSourceCode();
+    public function shouldGenerateSourceCode(): bool;
```
Comment (jankal) [inline src/Contracts/Configuration/ConfigurationInterface.php]: Typed `bool` return but no docblock.

File: src/Contracts/Console/Helper/ProgressBarInterface.php [hunk 17]
```diff
@@ -1,18 +1,12 @@
...
 namespace ApiGen\Contracts\Console\Helper;
 
 interface ProgressBarInterface
 {
 
-    /**
-     * @param int $maximum
-     */
-    public function init($maximum = 1);
+    public function init(int $maximum = 1): void;
```
Comment (jankal) [inline src/Contracts/Console/Helper/ProgressBarInterface.php]: This is a interface. Why does this afterwards implemented method not have any docblock about its typed param?

File: src/Contracts/Console/Helper/ProgressBarInterface.php [hunk 18]
```diff
@@ -1,18 +1,12 @@
...
 
 
-    /**
-     * @param int $increment
-     */
-    public function increment($steps = 1);
+    public function increment(int $steps = 1): void;
```
Comment (jankal) [inline src/Contracts/Console/Helper/ProgressBarInterface.php]: This is a interface. Why does this afterwards implemented method not have any docblock about its typed param?

File: src/Contracts/Console/IO/IOInterface.php [hunk 19]
```diff
@@ -8,28 +8,11 @@
 interface IOInterface
 {
 
-    /**
-     * @return InputInterface
-     */
-    public function getInput();
+    public function getInput(): InputInterface;
```
Comment (jankal) [inline src/Contracts/Console/IO/IOInterface.php]: We do have a typed return here but no docblock.

File: src/Contracts/Console/IO/IOInterface.php [hunk 20]
```diff
@@ -8,28 +8,11 @@
...
 
 
-    /**
-     * @return OutputInterface
-     */
-    public function getOutput();
+    public function getOutput(): OutputInterface;
```
Comment (jankal) [inline src/Contracts/Console/IO/IOInterface.php]: We do have a typed return here but no docblock.

File: src/Contracts/Console/IO/IOInterface.php [hunk 21]
```diff
@@ -8,28 +8,11 @@
...
-    public function writeln($message);
-
-
-    /**
-     * @param string $question
-     * @param NULL|string $default
-     * @return string
-     */
-    public function ask($question, $default = null);
+    public function writeln(string $message): void;
```
Comment (jankal) [inline src/Contracts/Console/IO/IOInterface.php]: We do have a typed param here but no docblock.

File: src/Contracts/Console/Input/DefaultInputDefinitionFactoryInterface.php [hunk 22]
```diff
@@ -1,14 +1,10 @@
...
 use Symfony\Component\Console\Input\InputDefinition;
 
 interface DefaultInputDefinitionFactoryInterface
 {
-
-    /**
-     * @return InputDefinition
-     */
-    public function create();
+    public function create(): InputDefinition;
```
Comment (jankal) [inline src/Contracts/Console/Input/DefaultInputDefinitionFactoryInterface.php]: We do have a typed return here but no docblock.

File: src/Contracts/Generator/GeneratorQueueInterface.php [hunk 23]
```diff
@@ -1,20 +1,13 @@
-<?php
+<?php declare(strict_types=1);
 
 namespace ApiGen\Contracts\Generator;
 
 use ApiGen\Contracts\Generator\TemplateGenerators\TemplateGeneratorInterface;
 
 interface GeneratorQueueInterface
 {
+    public function addToQueue(TemplateGeneratorInterface $templateGenerator): void;
```
Comment (jankal) [inline src/Contracts/Generator/GeneratorQueueInterface.php]: We do have a typed param here but no docblock.

File: src/Contracts/Generator/Resolvers/ElementResolverInterface.php [hunk 24]
```diff
@@ -1,44 +1,32 @@
...
 interface ElementResolverInterface
 {
-
-    /**
-     * @param string $className
-     * @param string $namespace
-     * @return ClassReflectionInterface|bool
-     */
-    public function getClass($className, $namespace = '');
+    public function getClass(string $className, string $namespace = ''): ?ClassReflectionInterface;
```
Comment (jankal) [inline src/Contracts/Generator/Resolvers/ElementResolverInterface.php]: We do have a typed return and two typed params here but no docblock.

This also is an interface so the implementations won't have any docs neither.

File: src/Contracts/Generator/Resolvers/ElementResolverInterface.php [hunk 25]
```diff
@@ -1,44 +1,32 @@
...
 
 
     /**
-     * @param string $name
-     * @param string $namespace
-     * @return FunctionReflectionInterface|NULL
+     * @return FunctionReflectionInterface|MethodReflectionInterface|null
      */
-    public function getFunction($name, $namespace = '');
+    public function getFunction(string $name, string $namespace = '');
```
Comment (jankal) [inline src/Contracts/Generator/Resolvers/ElementResolverInterface.php]: We do have a docblock here which states the return type but there's nothing about the params inside that docblock.

File: src/Contracts/Generator/Resolvers/ElementResolverInterface.php [hunk 26]
```diff
@@ -1,44 +1,32 @@
...
+    public function getClass(string $className, string $namespace = ''): ?ClassReflectionInterface;
 
 
-    /**
-     * @param string $className
-     * @param string $namespace
-     * @return ConstantReflectionInterface|bool
-     */
-    public function getConstant($className, $namespace = '');
+    public function getConstant(string $className, string $namespace = ''): ?ConstantReflectionInterface;
```
Comment (jankal) [inline src/Contracts/Generator/Resolvers/ElementResolverInterface.php]: We do have a typed return and two typed params here but no docblock.

This also is an interface so the implementations won't have any docs neither.

File: src/Contracts/Generator/StepCounterInterface.php [hunk 27]
```diff
@@ -1,12 +1,9 @@
...
 namespace ApiGen\Contracts\Generator;
 
 interface StepCounterInterface
 {
 
-    /**
-     * @return int
-     */
-    public function getStepCount();
+    public function getStepCount(): int;
```
Comment (jankal) [inline src/Contracts/Generator/StepCounterInterface.php]: This method inside an interface doesn't have any docs for the return type of `int` here.

File: src/Contracts/Generator/TemplateGenerators/ConditionalTemplateGeneratorInterface.php [hunk 28]
```diff
@@ -1,12 +1,9 @@
...
 namespace ApiGen\Contracts\Generator\TemplateGenerators;
 
 interface ConditionalTemplateGeneratorInterface extends TemplateGeneratorInterface
 {
 
-    /**
-     * @return bool
-     */
-    public function isAllowed();
+    public function isAllowed(): bool;
```
Comment (jankal) [inline src/Contracts/Generator/TemplateGenerators/ConditionalTemplateGeneratorInterface.php]: This method of this interface doesn't have any docs for it's return type.

File: src/Contracts/Parser/Broker/BackendInterface.php [hunk 29]
```diff
@@ -1,35 +1,28 @@
...
-
     /**
-     * Returns all classes from all namespaces.
-     *
      * @param int $type Returned class types (multiple values may be OR-ed).
      * @return ClassReflectionInterface[]
      */
-    public function getClasses($type);
-
+    public function getClasses($type = Memory::TOKENIZED_CLASSES);
```
Comment (jankal) [inline src/Contracts/Parser/Broker/BackendInterface.php]: According to docblock has typed return but I can't find any fixation of this in form of hinting this as a PHP return type with `: array`.

File: src/Contracts/Parser/FileProcessingExceptionInterface.php [hunk 30]
```diff
@@ -1,12 +1,8 @@
-<?php
+<?php declare(strict_types=1);
 
 namespace ApiGen\Contracts\Parser;
 
 interface FileProcessingExceptionInterface
```
Comment (jankal) [inline src/Contracts/Parser/FileProcessingExceptionInterface.php]: Is this interface/contract still needed even if it doesn't provide/ensure any contracted behaviour or methods?

File: src/Contracts/Parser/ParserInterface.php [hunk 31]
```diff
@@ -1,25 +1,15 @@
...
-    public function parse(array $files);
-
-
-    /**
-     * Get list of error found while parsing the code.
-     *
-     * @return array
      */
-    public function getErrors();
+    public function parse(array $files): ParserStorageInterface;
```
Comment (jankal) [inline src/Contracts/Parser/ParserInterface.php]: There is a docblock for this method but the docblock doesn't specify the return of this method.

File: src/Contracts/Parser/ParserStorageInterface.php [hunk 32]
```diff
@@ -1,75 +1,75 @@
...
 interface ParserStorageInterface
 {
-
     /**
-     * @param string $type
-     * @return ArrayObject
+     * @return mixed[]
      */
-    public function getElementsByType($type);
+    public function getElementsByType(string $type): array;
```
Comment (jankal) [inline src/Contracts/Parser/ParserStorageInterface.php]: Param `string $type` is not documented by the existing docblock.

File: src/Contracts/Parser/Reflection/AbstractFunctionMethodReflectionInterface.php [hunk 33]
```diff
@@ -1,25 +1,21 @@
...
 namespace ApiGen\Contracts\Parser\Reflection;
 
 interface AbstractFunctionMethodReflectionInterface extends ElementReflectionInterface
 {
-
-    /**
-     * @return bool
-     */
-    public function returnsReference();
+    public function returnsReference(): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/AbstractFunctionMethodReflectionInterface.php]: Oops. A return of `bool` is strictly ensured but there's nothing about that in the not existing docblock for this method.

File: src/Contracts/Parser/Reflection/Behavior/InClassInterface.php [hunk 34]
```diff
@@ -7,32 +7,20 @@
...
-     * @return ClassReflectionInterface
-     */
-    public function getDeclaringClass();
-
-
-    /**
-     * @return string
-     */
-    public function getPackageName();
+    public function getDeclaringClass(): ?ClassReflectionInterface;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/Behavior/InClassInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/Behavior/InClassInterface.php [hunk 35]
```diff
@@ -7,32 +7,20 @@
...
 
 
-    /**
-     * @return string
-     */
-    public function getNamespaceName();
+    public function getNamespaceName(): string;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/Behavior/InClassInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/Behavior/InNamespaceInterface.php [hunk 36]
```diff
@@ -1,19 +1,15 @@
...
 
 interface InNamespaceInterface
 {
 
-    /**
-     * @deprecated To be removed with ApiGen\ElementParser
-     * @return string
-     */
-    public function getDeclaringClassName();
+    public function getDeclaringClassName(): string;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/Behavior/InNamespaceInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/Behavior/InTraitInterface.php [hunk 37]
```diff
@@ -7,14 +7,8 @@
 interface InTraitInterface extends InNamespaceInterface
 {
 
-    /**
-     * @return ClassReflectionInterface
-     */
-    public function getDeclaringTrait();
+    public function getDeclaringTrait(): ?ClassReflectionInterface;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/Behavior/InTraitInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/Behavior/InTraitInterface.php [hunk 38]
```diff
@@ -7,14 +7,8 @@
...
 
 
-    /**
-     * @return string|NULL
-     */
-    public function getDeclaringTraitName();
+    public function getDeclaringTraitName(): string;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/Behavior/InTraitInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/Behavior/LinedInterface.php [hunk 39]
```diff
@@ -1,18 +1,12 @@
...
 namespace ApiGen\Contracts\Parser\Reflection\Behavior;
 
 interface LinedInterface
 {
 
-    /**
-     * @return int
-     */
-    public function getStartLine();
+    public function getStartLine(): int;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/Behavior/LinedInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/Behavior/LinedInterface.php [hunk 40]
```diff
@@ -1,18 +1,12 @@
...
 
 
-    /**
-     * @return int
-     */
-    public function getEndLine();
+    public function getEndLine(): int;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/Behavior/LinedInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/Behavior/NamedInterface.php [hunk 41]
```diff
@@ -1,19 +1,12 @@
...
 namespace ApiGen\Contracts\Parser\Reflection\Behavior;
 
 interface NamedInterface
 {
 
-    /**
-     * @return string
-     */
-    public function getName();
+    public function getName(): string;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/Behavior/NamedInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/Behavior/NamedInterface.php [hunk 42]
```diff
@@ -1,19 +1,12 @@
...
 
 
-    /**
-     * @return string
-     * @todo dunno if this belongs here
-     */
-    public function getPrettyName();
+    public function getPrettyName(): string;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/Behavior/NamedInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 43]
```diff
@@ -10,398 +10,299 @@
...
-     * @return bool
-     */
-    public function isDocumented();
-
-
-    /**
-     * @return bool
-     */
-    public function isValid();
+    public function isDocumented(): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 44]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return ClassReflectionInterface|NULL
-     */
-    public function getParentClass();
+    public function getParentClass(): ?ClassReflectionInterface;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 45]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return string
-     */
-    public function getParentClassName();
+    public function getParentClassName(): ?string;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 46]
```diff
@@ -10,398 +10,299 @@
...
-    public function getIndirectUsers();
+    public function getIndirectUsers(): array;
 
 
-    /**
-     * @param string $name
-     * @return bool
-     */
-    public function implementsInterface($name);
+    public function implementsInterface(string $name): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 47]
```diff
@@ -10,398 +10,299 @@
...
-    public function getOwnMagicMethods();
+    public function getOwnMagicMethods(): array;
 
 
-    /**
-     * @param string $name
-     * @return MethodReflectionInterface
-     */
-    public function getMethod($name);
+    public function getMethod(string $name): MethodReflectionInterface;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and arg not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 48]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @param string $name
-     * @return bool
-     */
-    public function hasMethod($name);
+    public function hasMethod(string $name): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and arg not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 49]
```diff
@@ -10,398 +10,299 @@
...
-    public function getInheritedConstants();
+    public function getInheritedConstants(): array;
 
 
-    /**
-     * @param string $name
-     * @return bool
-     */
-    public function hasConstant($name);
+    public function hasConstant(string $name): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and arg not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 50]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @param string $name
-     * @return bool
-     */
-    public function hasOwnConstant($name);
+    public function hasOwnConstant(string $name): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and arg not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 51]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @param string $name
-     * @return ClassConstantReflectionInterface
-     */
-    public function getConstant($name);
+    public function getConstant(string $name): ConstantReflectionInterface;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and arg not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 52]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @param string $name
-     * @return bool
-     */
-    public function getOwnConstant($name);
+    public function getOwnConstant(string $name): ConstantReflectionInterface;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and arg not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 53]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @param string $name
-     * @return ClassConstantReflectionInterface
-     */
-    public function getConstantReflection($name);
+    public function getDocComment(): string;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 54]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return string
-     */
-    public function getDocComment();
+    public function isVisibilityLevelPublic(): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 55]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return bool
-     */
-    public function isVisibilityLevelPublic();
+    public function getVisibilityLevel(): int;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 56]
```diff
@@ -10,398 +10,299 @@
...
-     * @return int
-     */
-    public function getVisibilityLevel();
-
-
-    /**
-     * @return ReflectionFactoryInterface
-     */
-    public function getReflectionFactory();
+    public function getReflectionFactory(): ReflectionFactoryInterface;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 57]
```diff
@@ -10,398 +10,299 @@
...
-    public function getUsedMagicProperties();
+    public function getUsedMagicProperties(): array;
 
 
-    /**
-     * @param string $name
-     * @return PropertyReflectionInterface
-     */
-    public function getProperty($name);
+    public function getProperty(string $name): PropertyReflectionInterface;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and param not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 58]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @param string $name
-     * @return bool
-     */
-    public function hasProperty($name);
+    public function hasProperty(string $name): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and param not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 59]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @param string $name
-     * @return bool
-     */
-    public function usesTrait($name);
+    public function usesTrait(string $name): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and param not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 60]
```diff
@@ -10,398 +10,299 @@
...
      */
-    public function getParsedClasses();
+    public function getParsedClasses(): array;
 
 
-    /**
-     * @return int
-     */
-    public function getStartPosition();
+    public function getStartPosition(): int;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 61]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return int
-     */
-    public function getEndPosition();
+    public function getEndPosition(): int;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 62]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return bool
-     */
-    public function isAbstract();
+    public function isAbstract(): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 63]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return bool
-     */
-    public function isFinal();
+    public function isFinal(): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 64]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return bool
-     */
-    public function isInterface();
+    public function isInterface(): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 65]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return bool
-     */
-    public function isException();
+    public function isException(): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 66]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @return bool
-     */
-    public function isTrait();
+    public function isTrait(): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ClassReflectionInterface.php [hunk 67]
```diff
@@ -10,398 +10,299 @@
...
 
 
-    /**
-     * @param string $class
-     * @return bool
-     */
-    public function isSubclassOf($class);
+    public function isSubclassOf(string $class): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ClassReflectionInterface.php]: Typed return and arg not documented here.

File: src/Contracts/Parser/Reflection/ConstantReflectionInterface.php [hunk 68]
```diff
@@ -13,10 +13,7 @@
     LinedInterface
 {
 
-    /**
-     * @return string
-     */
-    public function getTypeHint();
+    public function getTypeHint(): string;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ConstantReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ConstantReflectionInterface.php [hunk 69]
```diff
@@ -25,8 +22,5 @@ public function getTypeHint();
     public function getValue();
 
 
-    /**
-     * @return string
-     */
-    public function getValueDefinition();
+    public function getValueDefinition(): string;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ConstantReflectionInterface.php]: Typed return not documented here.

File: src/Contracts/Parser/Reflection/ElementReflectionInterface.php [hunk 70]
```diff
@@ -1,69 +1,24 @@
...
 
 namespace ApiGen\Contracts\Parser\Reflection;
 
-use ApiGen\Contracts\Parser\Reflection\Behavior\LinedInterface;
 use ApiGen\Contracts\Parser\Reflection\Behavior\NamedInterface;
-use TokenReflection\Exception\BaseException;
 
 interface ElementReflectionInterface extends NamedInterface
 {
+    public function isMain(): bool;
```
Comment (jankal) [inline src/Contracts/Parser/Reflection/ElementReflectionInterface.php]: Typed return not documented here. But not sure if it needs to be cause of doc inheritance.

File: src/Console/Command/AbstractCommand.php [hunk 71]
```diff
@@ -1,27 +1,26 @@
...
 abstract class AbstractCommand extends Command
 {
-
     /**
-     * {@inheritdoc}
+     * @param string $name
+     * @param ?string $shortcut
+     * @param ?int $mode
+     * @param string $description
+     * @param ?mixed $default
```
Comment (TomasVotruba) [inline src/Console/Command/AbstractCommand.php]: You are right. This is how I imagined the future. Will fix

File: src/Configuration/ConfigurationOptionsResolver.php [hunk 72]
```diff
@@ -1,52 +1,45 @@
...
+final class ConfigurationOptionsResolver
 {
-
-    const AL_PROTECTED = 'protected';
-    const AL_PRIVATE = 'private';
-    const AL_PUBLIC = 'public';
-    const DEFAULT_THEME = 'default';
+    public const AL_PROTECTED = 'protected';
+    public const AL_PRIVATE = 'private';
+    public const AL_PUBLIC = 'public';
```
Comment (TomasVotruba) [inline src/Configuration/ConfigurationOptionsResolver.php]: Good catch, will add

File: src/Contracts/Configuration/ConfigurationInterface.php (hunk 8 above)
Comment (TomasVotruba) [inline src/Contracts/Configuration/ConfigurationInterface.php]: Good point, I will add checker for that to CodingStandard

**Pull request merged**